### PR TITLE
Notify systemd when Elasticsearch is ready

### PR DESCRIPTION
**Pull request by jasontedor**
Today our systemd service defaults to a service type of simple. This means that systemd assumes Elasticsearch is ready as soon as the ExecStart (bin/elasticsearch) process is forked off. This means that the service appears ready long before it actually is, so before it is ready to receive requests. It also means that services that want to depend on Elasticsearch being ready to start can not as there is not a reliable mechanism to determine this. This commit changes the service type to notify. This requires that Elasticsearch sends a notification message via libsystemd sd_notify method. This commit does that by using JNA to invoke this native method. Additionally, we use this integration to also notify systemd when we are stopping.

### CODE REVIEW

**Review by alpar-t — COMMENTED** [review]
Would it be worth adding a packaging test to test this integration ?

File: modules/systemd/src/main/java/org/elasticsearch/systemd/Libsystemd.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import java.security.PrivilegedAction;
+
+/**
+ * Provides access to the native method sd_notify from libsystemd.
+ */
+class Libsystemd {
+
+    static {
+        AccessController.doPrivileged((PrivilegedAction<Object>) () -> {
+            Native.register(Libsystemd.class, "libsystemd.so.0");
```
Comment (atorok) [inline modules/systemd/src/main/java/org/elasticsearch/systemd/Libsystemd.java]: Won't this fail with the  packaged distributions on OS-es that don't support systemd, or does everything in our support matrix support it ?

Comment (jasontedor) [reply modules/systemd/src/main/java/org/elasticsearch/systemd/Libsystemd.java]: We don’t try to load the native library if the environment variable ES_SD_NOTIFY is not set to “true” and in this PR we do that only in the systemd service integration (see the unit file).

**Comment by jasontedor** [comment]
> Would it be worth adding a packaging test to test this integration ?

I considered the existing packaging tests sufficient, since if this integration is broken due to the service type being set to `notfiy` service start calls would never return.

I’ll look more carefully if there is deeper testing we can do but I considered the existing packaging tests sufficient.

File: modules/systemd/src/main/java/org/elasticsearch/systemd/SystemdPlugin.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+            return;
+        }
+        if (Boolean.TRUE.toString().equals(esSDNotify) == false && Boolean.FALSE.toString().equals(esSDNotify) == false) {
+            throw new RuntimeException("ES_SD_NOTIFY set to unexpected value [" + esSDNotify + "]");
+        }
+        enabled = "true".equals(esSDNotify);
+    }
+
+    int sd_notify(@SuppressWarnings("SameParameterValue") final int unset_environment, final String state) {
+        return Libsystemd.sd_notify(0, state);
```
Comment (jasontedor) [inline modules/systemd/src/main/java/org/elasticsearch/systemd/SystemdPlugin.java]:
```suggestion
        return Libsystemd.sd_notify(unset_environment, state);
```

**Comment by jasontedor** [comment]
@elasticmachine update branch

File: server/src/main/resources/org/elasticsearch/bootstrap/security.policy [hunk 3]
```diff
@@ -24,6 +24,10 @@
 //// SecurityManager impl:
 //// Must have all permissions to properly perform access checks
 
+grant codeBase "${codebase.jna}" {
```
Comment (rjernst) [inline server/src/main/resources/org/elasticsearch/bootstrap/security.policy]: Can this go below the "Very speciall jar permissions" comment please, with the rest of the jar specific grants? (yes the SM here is a jar specific grant, but the comment above is specific to secure SM)

**Review by rjernst — APPROVED** [review]
LGTM, one minor comment. Thanks for making this only included in the packages and not archives.

**Pull request merged**